### DISCUSSION TITLE
Should ValueElement be `Generic[V]` ?

### DISCUSSION THREAD

**Discussion by petergaultney** [zauberzeug/nicegui#2677] [Ideas / Feature Requests]
NiceGUI has the concept of a ValueElement, which contains a `value` attribute (Bindable Property) that can be set and retrieved.

For the subset of these that I've taken a look at, they have a 'correct' type of the value. For instance, `Input` has `value: str`, whereas `number` has `value: float`

At the moment, the code types these as `Any` because the parent mixin (ValueElement) is not represented as Generic. But it seems like that would be possible, and not too hard to do? And then we could change to `set_value(value: V)` (instead of `Any`), and more NiceGUI-using code could take advantage of types.

**Comment by falkoschindler** [comment]
Hi @petergaultney, that's a great idea! It always feels a bit odd that the `value` can be of `Any` type.

It's hard to foresee all consequences though. E.g. there's the `ValueChangeEventArguments` class which cnsequently should also be generic. But we can start with the `ValueElement`, let mypy and pylint guide our way and see how it goes.

Would you like to give it a go? 🙂

**Reply by petergaultney** [reply]
let's see how my other PR goes (builder pattern for event handlers), and then I will give this one a try. 

I agree that there are some unpredictable consequences of trying to refactor types at this scale. :)  Of course, for some users it may also start causing them static typing errors if we improve this - but for those who use static typing, I think they will mostly be grateful even if there is some refactoring they need to do.

**Reply by jermatic** [reply]
theres a big shift in python 3.12 where generics become first class citizens so..... id just keep that in mind when moving forward

**Comment by evnchn** [comment]
Got it working in https://github.com/zauberzeug/nicegui/pull/5785

**Reply by falkoschindler** [reply]
I'm closing this feature request as resolved in version 3.11 ✔️
